bingo/local_optimizers/continuous_local_opt.py

```diff
@@ -162,104 +56,24 @@ def eval_count(self):
-        if algorithm in ROOT_SET and not hasattr(fitness_function,
-                                                 'evaluate_fitness_vector'):
-            raise TypeError("{} requires VectorBasedFunction\
-                            as a fitness function".format(algorithm))
-
-    def _optimize_params(self, individual):
-        num_params = individual.get_number_local_optimization_params()
-        c_0 = np.random.uniform(*self._param_init_bounds, num_params)
-        params = self._run_algorithm_for_optimization(
-            self._sub_routine_for_fit_function, individual, c_0)
-        individual.set_local_optimization_params(params)
-
-    def _sub_routine_for_fit_function(self, params, individual):
-        individual.set_local_optimization_params(params)
-
-        if self._algorithm in ROOT_SET:
-            return self._fitness_function.evaluate_fitness_vector(individual)
-        return self._fitness_function(individual)
-
-    def _run_algorithm_for_optimization(self, sub_routine, individual, params):
-        if self._algorithm in ROOT_SET:
-            if isinstance(self._fitness_function, VectorGradientMixin) \
-                    and self._algorithm in JACOBIAN_SET:
-                optimize_result = optimize.root(
-                        sub_routine, params,
-                        args=(individual, ),
-                        jac=lambda x, indv: self._fitness_function.get_fitness_vector_and_jacobian(indv)[1],  # pylint: disable=line-too-long
-                        method=self._algorithm,
-                        **self._optimization_options)
-            else:
-                optimize_result = optimize.root(sub_routine, params,
-                                                args=(individual),
-                                                method=self._algorithm,
-                                                **self._optimization_options)
-        else:
-            if isinstance(self._fitness_function, GradientMixin) \
-                    and self._algorithm in JACOBIAN_SET:
-                optimize_result = optimize.minimize(
-                        sub_routine, params,
-                        args=(individual, ),
-                        jac=lambda x, indv: self._fitness_function.get_fitness_and_gradient(indv)[1],  # pylint: disable=line-too-long
-                        method=self._algorithm,
-                        **self._optimization_options)
-            else:
-                optimize_result = optimize.minimize(sub_routine, params,
-                                                    args=(individual, ),
-                                                    method=self._algorithm,
-                                                    **self._optimization_options)  # pylint: disable=line-too-long
-        return optimize_result.x
-
-    def _evaluate_fitness(self, individual):
+            self.optimizer(individual)
         return self._fitness_function(individual)
 
 
```

Comment:
I think we could combine the interface into ```Chromosome```. If that were the case, I would want to add a ```params``` property to ```Chromosome```. However, to me, the interface suggests there exists a case where we don't want to optimize every parameter. In this case, it would be nice to have the interface and I would see merit to not combining it with ```Chromosome``` since it's possible not every ```Chromosome``` with parameters would care about optimization.

An alternative that maybe you were referencing is we could include the interface in ```Chromosome``` but set it up to the normal case where every parameter would be optimized. This way, the case where not every parameter is optimized can still be configured via ```get_number_local_optimization_params``` and ```set_local_optimization_params```, but not every ```Chromosome``` has to implement them. If we want to go with this approach, should we keep ```needs_local_optimization``` abstract? Maybe it can turn into a property (e.g. ```needs_opt``` or similar)?

Reply:
Decided on renaming ```ChromosomeInterface``` to ```LocalOptimizationInterface``` and moving it to its own file. We also decided that ```Chromosome``` should implement ```LocalOptimizationInterface``` but raise a ```NotImplementedError``` in the interface methods by default.

I couldn't think of a significant difference between ```needs_local_optimization``` and a ```needs_opt``` property, so I left the ```needs_local_optimization``` method in the interface. Let me know if you can think of any benefits to the property approach.